src/main/java/net/dv8tion/jda/api/utils/data/etf/ExTermEncoder.java

```diff
@@ -0,0 +1,147 @@
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import java.util.Map;
+
+import static net.dv8tion.jda.api.utils.data.etf.ExTermTag.*;
+
+public class ExTermEncoder
+{
+
+    public static ByteBuffer pack(Object data)
+    {
+        ByteBuffer buffer = ByteBuffer.allocate(1024);
+        buffer.put((byte) 131);
+
+        return pack(buffer, data).flip();
+    }
+
+    @SuppressWarnings("unchecked")
+    public static ByteBuffer pack(ByteBuffer buffer, Object value)
+    {
+        if (value instanceof String)
+            return packBinary(buffer, (String) value);
+        if (value instanceof Map)
+            return packMap(buffer, (Map<String, Object>) value);
+        if (value instanceof List)
+            return packList(buffer, (List<Object>) value);
+        if (value instanceof Integer)
+            return packInt(buffer, (int) value);
+        if (value instanceof Long)
+            return packLong(buffer, (long) value);
+        if (value instanceof Boolean)
+            return packAtom(buffer, String.valueOf(value));
+        if (value == null)
+            return packAtom(buffer, "nil");
+
+        throw new UnsupportedOperationException();
+    }
+
+    private static ByteBuffer realloc(ByteBuffer buffer, int length)
+    {
+        if (buffer.remaining() >= length)
+            return buffer;
+
+        ByteBuffer allocated = ByteBuffer.allocate((buffer.position() + length) * 2);
+        allocated.put(buffer.flip());
+        return allocated;
+    }
+
+    private static ByteBuffer packMap(ByteBuffer buffer, Map<String, Object> data)
+    {
+        buffer = realloc(buffer, data.size() + 5);
+        buffer.put(MAP);
+        buffer.putInt(data.size());
+
+        for (Map.Entry<String, Object> entry : data.entrySet())
+        {
+            buffer = packBinary(buffer, entry.getKey());
+            buffer = pack(buffer, entry.getValue());
+        }
+
+        return buffer;
+    }
+
+    private static ByteBuffer packList(ByteBuffer buffer, List<Object> data)
+    {
+        buffer = realloc(buffer, data.size() + 5);
+        buffer.put(LIST);
+        buffer.putInt(data.size());
+        for (Object element : data)
+        {
+            buffer = pack(buffer, element);
+        }
+        buffer.put(NIL);
+        return buffer;
+    }
+
+    private static ByteBuffer packBinary(ByteBuffer buffer, String value)
+    {
+        byte[] encoded = value.getBytes(StandardCharsets.UTF_8);
+        buffer = realloc(buffer, encoded.length + 5);
+        buffer.put(BINARY);
+        buffer.putInt(value.length());
+        buffer.put(encoded);
+        return buffer;
+    }
+
+    private static ByteBuffer packInt(ByteBuffer buffer, int value)
+    {
+        buffer = realloc(buffer, 5);
+        buffer.put(INT);
+        buffer.putInt(value);
+        return buffer;
+    }
+
+    private static ByteBuffer packLong(ByteBuffer buffer, long value)
+    {
+        byte sign = (byte) (value < 0 ? 1 : 0);
```

Comment:
I'm not sure if we should just treat all longs here as unsigned, discord doesn't really need signed longs for anything.